src/switchboard.rs

```diff
@@ -142,25 +145,30 @@ impl Switchboard {
         self.blockers_to_miscreants.disassociate(from, target);
     }
 
-    pub fn join_room(&mut self, session: Arc<Session>, room: RoomId) {
+    pub fn join_room(&mut self, session: Arc<Session>, user: UserId, room: RoomId) {
+        self.users_to_room.entry(user).or_insert(room.clone());
         self.occupants.entry(room).or_insert_with(Vec::new).push(session);
     }
 
-    pub fn leave_room(&mut self, session: &Session, room: RoomId) {
+    pub fn leave_room(&mut self, session: &Session, user: UserId, room: RoomId) {
         if let Entry::Occupied(mut cohabitators) = self.occupants.entry(room) {
             cohabitators.get_mut().retain(|x| x.as_ref() != session);
             if cohabitators.get().is_empty() {
                 cohabitators.remove_entry();
             }
         }
+
+        if let Entry::Occupied(room) = self.users_to_room.entry(user) {
+            room.remove_entry();
+        }
```

Comment:
This isn't checking to see whether it's the room passed in as an argument, but since we don't believe in the "same user, multiple rooms" case, it shouldn't matter for correctness. It's just confusing.